src/main/scala/format/BinaryFen.scala

```diff
@@ -0,0 +1,314 @@
+        Situation(
+          Board(
+            BBoard(
+              occupied = occupied,
+              white = white,
+              black = black,
+              pawns = pawns,
+              knights = knights,
+              bishops = bishops,
+              rooks = rooks,
+              queens = queens,
+              kings = kings
+            ),
+            History(
+              lastMove = epMove,
+              checkCount = checkCount,
+              castles =
+                maximumCastles(unmovedRooks = unmovedRooks, white = white, black = black, kings = kings),
+              unmovedRooks = unmovedRooks,
+              halfMoveClock = halfMoveClock
+            ),
+            variant,
+            crazyData
+          ),
+          color = turn
+        ),
+        ply.fullMoveNumber
+      )
+
+  def write(input: Situation.AndFullMoveNumber): BinaryFen =
+    val builder = ArrayBuilder.ofByte()
+
+    val sit      = input.situation
+    val occupied = sit.board.occupied
+    writeLong(builder, occupied.value)
+
+    val unmovedRooks                 = minimumUnmovedRooks(sit.board)
+    val pawnPushedTo: Option[Square] = sit.enPassantSquare.map(_.xor(Square.A2))
+
+    def packPiece(sq: Square): Byte =
+      sit.board(sq) match
+        // Encoding from
+        // https://github.com/official-stockfish/nnue-pytorch/blob/2db3787d2e36f7142ea4d0e307b502dda4095cd9/lib/nnue_training_data_formats.h#L4607
+        case Some(Piece(_, Pawn)) if pawnPushedTo.contains(sq) => 12
+        case Some(Piece(White, Pawn))                          => 0
+        case Some(Piece(Black, Pawn))                          => 1
+        case Some(Piece(White, Knight))                        => 2
+        case Some(Piece(Black, Knight))                        => 3
+        case Some(Piece(White, Bishop))                        => 4
+        case Some(Piece(Black, Bishop))                        => 5
+        case Some(Piece(White, Rook))                          => if unmovedRooks.contains(sq) then 13 else 6
+        case Some(Piece(Black, Rook))                          => if unmovedRooks.contains(sq) then 14 else 7
+        case Some(Piece(White, Queen))                         => 8
+        case Some(Piece(Black, Queen))                         => 9
+        case Some(Piece(White, King))                          => 10
+        case Some(Piece(Black, King))                          => if sit.color.white then 11 else 15
+        case None                                              => 0 // unreachable
+
+    val it = occupied.iterator
+    while it.hasNext
+    do writeNibbles(builder, packPiece(it.next), if it.hasNext then packPiece(it.next) else 0)
+
+    val halfMoveClock = sit.history.halfMoveClock.value
+    val ply           = input.fullMoveNumber.ply(sit.color).value
+    val brokenTurn    = sit.color.black && sit.board(Black, King).isEmpty
+    val variantHeader = sit.variant match
+      case Standard | Chess960 | FromPosition => 0
```

Comment:
It seems We lose variant here, Should We give one extra byte for variant?

Reply:
good point. no reason not to, really